qiskit_experiments/calibration/experiments/rabi.py

```diff
@@ -0,0 +1,337 @@
+                linearly spaced between 0 and 2pi.
+            offset_guess: The y-axis offset which defaults to 0.5.
+            phase_guess: Phase of the oscillation which defaults to both 0 and pi.
+            amp_bounds: Bounds on the amplitude which default to (-1, 1).
+            freq_bounds: Bounds on the frequency which default to (0, inf).
+            offset_bounds: Bounds on the offset which default to (0,1).
+            phase_bounds: Bounds on the phase of the cosine which default to (-pi, pi).
+            plot: If True generate a plot of fitted data.
+            ax: Optional, matplotlib axis to add plot to.
+
+        Returns:
+            The analysis result with the fit and optional plots.
+
+        Raises:
+            QiskitError: If the fit fails.
+        """
+
+        meas_level = experiment_data.data(0)["metadata"]["meas_level"]
+        meas_return = experiment_data.data(0)["metadata"]["meas_return"]
+
+        # Pick a data processor.
+        if data_processor is None:
+            data_processor = get_to_signal_processor(meas_level=meas_level, meas_return=meas_return)
+            data_processor.train(experiment_data.data())
+
+        y_sigmas = np.array([data_processor(datum) for datum in experiment_data.data()])
+        min_y, max_y = min(y_sigmas[:, 0]), max(y_sigmas[:, 0])
+        ydata = (y_sigmas[:, 0] - min_y) / (max_y - min_y)
+
+        # Sigmas may be None and fitting will not work if any sigmas are exactly 0.
+        try:
+            sigmas = y_sigmas[:, 1] / (max_y - min_y)
+            if any(sigmas == 0.0):
+                sigmas = None
+
+        except TypeError:
+            sigmas = None
+
+        xdata = np.array([datum["metadata"]["xval"] for datum in experiment_data.data()])
+
+        # Perform fit
+        best_fit = None
+
+        def fit_fun(x, amplitude, frequency, phase, offset):
+            return amplitude * np.cos(frequency * x + phase) + offset
+
+        bounds = {
+            "amplitude": amp_bounds,
+            "frequency": freq_bounds,
+            "phase": phase_bounds,
+            "offset": offset_bounds,
+        }
+
+        # Guesses have two phases to catch signals that start at 1 or 0 at zero-amplitude.
+        guesses = []
+        for freq in freq_guesses:
+            if phase_guess is None:
+                guesses.append((freq, 0))
+                guesses.append((freq, np.pi))
+            else:
+                guesses.append((freq, phase_guess))
+
+        for guess in guesses:
+            freq_guess, phase_guess = guess[0], guess[1]
+            init = {
+                "amplitude": amp_guess,
+                "frequency": freq_guess,
+                "phase": phase_guess,
+                "offset": offset_guess,
+            }
+
+            try:
+                fit_result = curve_fit(fit_fun, xdata, ydata, init, sigmas, bounds)
+
+                if not best_fit:
+                    best_fit = fit_result
```

Comment:
I remember that I saw this idea, of trying several guesses and picking the best, in another PR of @eggerdj. I wonder if we could put this method, which tries and picks,  in a general location.

Reply:
Thinking about it, how will this experiment look like when the curve class is merged? Could we move this there? If yes then maybe this analysis class will only inherit from the curve class with a few adjustments of variables, as in the T1, Ramsey, and RB examples in the curve PR.

Reply:
@nkanazawa1989

Reply:
In curve analysis there is a method that generates fit condition `._setup_fitting`. If you return a list of conditions, the fitter is run multiple times and find the best result based on reduced chi squared value. So you don't need to write this logic in subclasses.

Reply:
Sounds good. Let's align once #46 is in.